### PR TITLE
Add communication-free Muon for FlexShard

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3502
* #3537
* #3603
* #3317
* #3239

Place each Muon-eligible 2D matrix whole on one rank via the Owned placement so Newton-Schulz runs locally on the owner after the backward reduce-to-owner -- no collective in optimizer.step(), and bit-exact with single-device Muon. Layers are balanced across ranks with greedy LPT; embeddings, LM head, and final norm stay Shard(0) + AdamW. Owned now composes with reshard_after_forward=True (broadcast ops tagged for activation-checkpoint recompute).

Adds example/muon.py (comm_free_muon_buckets, build_muon_param_groups, build_comm_free_muon_optimizers, CombinedOptimizer) and example/owned.py helpers (make_owned_placement_fn, assign_layer_owners_lpt).

Also adds GroupedMuon: batched Newton-Schulz over the leading dim of stacked weight matrices (>=3D, e.g. MoE grouped experts). build_comm_free_muon_optimizers routes 2D params to torch.optim.Muon and >=3D stacks to GroupedMuon; GroupedMuon matches running torch.optim.Muon on each 2D sub-matrix.

Tests: python -m pytest -q torchtitan/experiments/flex_shard/tests/test_flex_shard_muon.py

### CODE REVIEW

File: torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+This doc maps the placement design space for running the Muon optimizer on
+FlexShard-sharded transformer weights, organized by the tradeoff that defines it:
+**comm-free vs memory-balanced**. The two ends are duals —
+
+- **Comm-free `Owned`** (recommended default, implemented): the optimizer step
+  issues **zero collectives** and Newton-Schulz (NS) is **bit-exact**, at the cost
+  of **matrix-granular** memory balance (a rank owns whole matrices).
+- **Memory-balanced `GroupedRaggedShard`** (when comm-free is relaxed):
+  **byte-perfect** balance and no idle ranks, at the cost of an all-gather to
+  reconstruct each matrix for NS.
```
Comment (tianyu-l) [inline torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md]: what's the benefit of this over FSDP2?

File: torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+apply it on the owner and let the next forward's broadcast carry it.
+
+## Design
+
+### 1. Placement strategy
+
+Split the model into two regions:
+
+- **Transformer layers → `Owned`, one owner per layer (balanced by LPT).** Assign
+  each whole `layers.{i}.*` to a single owner rank and make that layer one Owned
```
Comment (tianyu-l) [inline torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md]: This cannot scale right? What if num_layers << world size?

File: torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+  bucket (one broadcast forward / one reduce backward, same granularity as today's
+  `transformer_bucket_specs`). The owner runs Muon on the layer's 2D matrices
+  (attention `wq/wk/wv/wo`, MLP `w1/w2/w3`) and AdamW on the layer's 1D norms —
```
Comment (tianyu-l) [inline torchtitan/experiments/flex_shard/muon_flex_shard_placement_strategies.md]: Where is the treatment that wqkv can be sharded by head_dim and still do comm-free Muon?

**Review by tianyu-l — CHANGES_REQUESTED** [review]
